subsys/jaxrs/src/main/java/org/commonjava/indy/bind/jaxrs/IndyRestMapperExceptionHandler.java

```diff
@@ -0,0 +1,56 @@
+package org.commonjava.indy.bind.jaxrs;
+
+
+import com.fasterxml.jackson.databind.jsonFormatVisitors.JsonObjectFormatVisitor;
+import org.commonjava.indy.IndyException;
+import org.commonjava.indy.IndyWorkflowException;
+import org.commonjava.indy.model.rest.IndyRestMapperResponse;
+import org.jboss.resteasy.client.exception.ResteasyHttpException;
+
+import javax.print.attribute.standard.Media;
+import javax.ws.rs.NotFoundException;
+import javax.ws.rs.ServiceUnavailableException;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.ext.ExceptionMapper;
+import javax.ws.rs.ext.Provider;
+import java.io.IOException;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+
+/*
+    Catch all Server Exceptions propagated back to client caller through HTTP REST calls and returning json object
+    representation of exception message.
+ */
+@Provider
+public class IndyRestMapperExceptionHandler implements ExceptionMapper<Exception> {
+
+
+    @Override
+    public Response toResponse(Exception exception) {
+        IndyRestMapperResponse irmpr = new IndyRestMapperResponse(exception.getMessage(), LocalDateTime.now());
+        Optional
+          .ofNullable(exception.getCause())
+          .ifPresent((exc) -> { irmpr.setCause(exc.getCause()); } );
+
+        if(exception instanceof IndyWorkflowException || exception instanceof IOException) {
+            return Response.status(Response.Status.BAD_REQUEST).entity(irmpr).type(MediaType.APPLICATION_JSON).build();
+        } else if(exception instanceof ServiceUnavailableException ) {
+            return Response.status(Response.Status.SERVICE_UNAVAILABLE).entity(irmpr).type(MediaType.APPLICATION_JSON).build();
+        } else if(exception instanceof NotFoundException) {
+//            return null;
+            return Response.status(Response.Status.NOT_FOUND).entity(irmpr).type(MediaType.APPLICATION_JSON).build();
+        }
+//        else if(exception instanceof RuntimeException) {
+//            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).entity(irmpr).type(MediaType.APPLICATION_JSON).build();
+//        }
+        else {
```

Comment:
This seems like it returns a 404 whenever it's an exception we don't understand? That seems misleading...in most cases where we hit this layer of code, we should be returning a 500 error I think. And I think we should always return the rest-mapper response if we have it...

I'm not sure we'd ever want to return a 404 from this method...am I missing something maybe?

Reply:
We also need special handling of InvocationTargetException, I think...that's a runtime exception but it has a different method call from getCause(), which will show the actual exception that happened. It's common when we use reflection, for instance. MissingMethodException is another one we have to deal with too, from Groovy scripts.

Reply:
That was my initial implementation but this "Return404DisableTimeoutForEnabledRepoTest" test is causing problems with that functionality.

Reply:
@jdcasey this is log from "/embedder/target/failsafe/ ... " :

===============
Exception in thread "Thread-13" java.lang.IllegalStateException: Singleton not set for fcf4a4b4-17ea-428f-884f-1a788a5c0ab0 => []
	at org.jboss.weld.bootstrap.api.helpers.RegistrySingletonProvider$RegistrySingleton.get(RegistrySingletonProvider.java:28)
	at org.jboss.weld.Container.instance(Container.java:63)
	at org.jboss.weld.bean.proxy.ClientProxyProvider.createClientProxy(ClientProxyProvider.java:200)
	at org.jboss.weld.bean.proxy.ClientProxyProvider.createClientProxy(ClientProxyProvider.java:196)
	at org.jboss.weld.bean.proxy.ClientProxyProvider.access$100(ClientProxyProvider.java:44)
	at org.jboss.weld.bean.proxy.ClientProxyProvider$CreateClientProxy.apply(ClientProxyProvider.java:52)
	at org.jboss.weld.bean.proxy.ClientProxyProvider$CreateClientProxy.apply(ClientProxyProvider.java:48)
	at org.jboss.weld.util.cache.ReentrantMapBackedComputingCache$1.apply(ReentrantMapBackedComputingCache.java:55)
	at org.jboss.weld.util.cache.ReentrantMapBackedComputingCache$1.apply(ReentrantMapBackedComputingCache.java:51)
	at org.jboss.weld.util.cache.ReentrantMapBackedComputingCache.getValue(ReentrantMapBackedComputingCache.java:64)
	at org.jboss.weld.util.cache.ReentrantMapBackedComputingCache.getCastValue(ReentrantMapBackedComputingCache.java:80)
	at org.jboss.weld.bean.proxy.ClientProxyProvider.getClientProxy(ClientProxyProvider.java:230)
	at org.jboss.weld.manager.BeanManagerImpl.getReference(BeanManagerImpl.java:780)
	at org.jboss.weld.bean.builtin.InstanceImpl.getBeanInstance(InstanceImpl.java:251)
	at org.jboss.weld.bean.builtin.InstanceImpl.access$200(InstanceImpl.java:73)
	at org.jboss.weld.bean.builtin.InstanceImpl$InstanceImplIterator.next(InstanceImpl.java:321)
	at org.commonjava.indy.action.IndyLifecycleManager.runShutdownActions(IndyLifecycleManager.java:251)
	at org.commonjava.indy.action.IndyLifecycleManager.stop(IndyLifecycleManager.java:172)
	at org.commonjava.indy.action.IndyLifecycleManager.lambda$createShutdownRunnable$0(IndyLifecycleManager.java:279)
	at java.lang.Thread.run(Thread.java:748)
=======================